--- a/basis_ffi.c
+++ b/basis_ffi.c
@@ -293,11 +293,6 @@
   #endif
 }
 
-void fficustom (unsigned char *c, long clen, unsigned char *a, long alen) {
-  assert(0 <= alen);
-  assert(0 <= clen);
-}
-
 // ---------------------------------------------------------------------------
 // Functions on doubles for the Double module
 // ---------------------------------------------------------------------------
@@ -376,6 +371,47 @@
     memcpy(a, d.bytes, sizeof d.bytes);
 }
 
+void fficustom (unsigned char *c, long clen, unsigned char *a, long alen) {
+  if (strcmp(c, "chdir") == 0) {
+    assert(1 <= alen);
+    a[0] = chdir(a) != 0;
+  } else if (strcmp(c, "getcwd") == 0) {
+    assert(1 <= alen);
+    a[0] = getcwd(a + 1, alen - 1) == NULL;
+  } else if (strcmp(c, "system") == 0) {
+    assert(2 <= alen);
+    assert(memchr(a, 0, alen) != NULL);
+    int ret = system((const char *)a);
+
+    // return convention:
+    // 0 < a[0] ==> no termination status for child
+    // a[0] = 0 ==> a[1] = termination status of child
+    if (ret == -1) {
+      a[0] = 1;
+      return;
+    }
+    if (!(WIFEXITED(ret))) {
+      a[0] = 2;
+      return;
+    }
+    a[0] = 0;
+    a[1] = WEXITSTATUS(ret);
+    return;
+  } else if (strcmp(c, "frexp") == 0) {
+    assert(alen >= 16);
+    double_bytes d;
+    int_bytes i;
+    int exp;
+    memcpy(d.bytes, a, sizeof d.bytes);
+    d.num = frexp(d.num, &exp);
+    i.num = exp;  // int -> int64_t
+    memcpy(a, d.bytes, sizeof d.bytes);
+    memcpy(&a[8], i.bytes, sizeof i.bytes);
+  } else {
+    return;  // Ignore unknown command
+  }
+}
+
 void cml_clear() {
   __builtin___clear_cache(&cake_codebuffer_begin, &cake_codebuffer_end);
 }
